Цель работы:

Осуществить предварительную обработку данных csv-файла, выявить и устранить проблемы в этих данных.


# Загрузка набора данных
## Описание предметной области
Вариант №5
<br>Набор данных visits.csv <br> Атрибуты:
<br>User_Id — уникальный идентификатор пользователя
<br>Region — страна пользователя
<br>Device — устройство пользователя
<br>Channel — идентификатор рекламного источника, из которого пришел пользователь
<br>Session_Start — дата и время начала сессии
<br>Session_End — дата и время окончания сессии

---
## 1. Чтение файла (набора данных)
Импорт библиотек, чтение файла с помощью pandas

In [21]:
import pandas as pd
df = pd.read_csv("../data/visits.csv", sep =';')

---
## 2. Обзор данных
 2.1 Вывод первых 20 строк с помощью метода head

In [22]:
df.head(20)

User_Id         Region   Device       Channel     Session_Start  \
0   9,81449E+11  United States   iPhone       organic   01.05.2019 2:36   
1   2,78966E+11  United States   iPhone       organic   01.05.2019 4:46   
2   5,90706E+11  United States      Mac       organic  01.05.2019 14:09   
3   3,26434E+11  United States  Android        TipTop   01.05.2019 0:29   
4   3,49774E+11  United States      Mac       organic   01.05.2019 3:33   
5   43958116050  United States  Android       organic   01.05.2019 9:03   
6   1,85365E+11  United States   iPhone       organic   01.05.2019 9:37   
7   1,00971E+11  United States      Mac        TipTop   01.05.2019 4:39   
8   3,70456E+11  United States   iPhone       organic  01.05.2019 14:44   
9   1,41838E+11  United States      Mac      FaceBoom   01.05.2019 6:20   
10  9,24309E+11  United States   iPhone       organic  01.05.2019 20:03   
11  7,74938E+11  United States   iPhone  MediaTornado  01.05.2019 14:01   
12  2,44879E+11  United States   iPhone  MediaTornado   01.05.2019 1:59   
13  1,57877E+11  United States      Mac       organic  01.05.2019 10:25   
14  5,24629E+11  United States      Mac       organic  01.05.2019 17:06   
15  1,36227E+11  United States       PC       organic  01.05.2019 19:34   
16  6,84979E+11  United States   iPhone        TipTop   01.05.2019 1:51   
17  9,05776E+11  United States   iPhone       organic  01.05.2019 14:26   
18  15174034813  United States   iPhone      FaceBoom  01.05.2019 15:34   
19  1,88246E+11  United States  Android      FaceBoom  01.05.2019 23:09   

         SESSION_End  
0    01.05.2019 2:45  
1    01.05.2019 4:47  
2   01.05.2019 15:32  
3    01.05.2019 0:54  
4    01.05.2019 3:57  
5   01.05.2019 10:08  
6   01.05.2019 10:00  
7    01.05.2019 4:57  
8   01.05.2019 15:41  
9    01.05.2019 6:54  
10  01.05.2019 20:19  
11  01.05.2019 14:29  
12   01.05.2019 2:21  
13  01.05.2019 10:35  
14  01.05.2019 18:05  
15  01.05.2019 19:39  
16   01.05.2019 2:12  
17  01.05.2019 14:35  
18  01.05.2019 15:56  
19  01.05.2019 23:21

---
 2.2 Оценка данных с помощью метода info

In [23]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 957 entries, 0 to 956
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   User_Id        957 non-null    object
 1   Region         956 non-null    object
 2   Device         955 non-null    object
 3   Channel        957 non-null    object
 4   Session_Start  955 non-null    object
 5   SESSION_End    955 non-null    object
dtypes: object(6)
memory usage: 45.0+ KB


---
2.3 Оценка данных с помощью метода describe.

In [24]:
df.describe()

User_Id         Region  Device  Channel     Session_Start  \
count           957            956     955      957               955   
unique          865              2       6        4               820   
top     1,79085E+11  United States  iPhone  organic  02.05.2019 20:16   
freq              4            955     421      612                 4   

             SESSION_End  
count                955  
unique               836  
top     02.05.2019 22:48  
freq                   3

---
Была произведена оценка данных методами describe и info. В датасете 957 строк и 6 столбцов, тип всех столбцов строковый, пропуски данных в столбцах Region(1), Device(2), Session_Start(2), Session_End(2).
***


2.4 Оценка названий столбцов

In [25]:
df.columns

Index(['User_Id', 'Region', 'Device', 'Channel', 'Session_Start',
       'SESSION_End'],
      dtype='object')

2.5 Анализ и коррекция названий столбцов
На экран были выведены названия столбцов, ошибка есть только в название последнего столбца SESSION_End, слово целиком написано в верхнем регистре, для удобства приведем все к единому виду.

In [26]:
df.columns = df.columns.str.lower()
df.columns

Index(['user_id', 'region', 'device', 'channel', 'session_start',
       'session_end'],
      dtype='object')

---
### 3. Проверка пропусков

In [27]:
df.isna().sum()

user_id          0
region           1
device           2
channel          0
session_start    2
session_end      2
dtype: int64

In [28]:
df['region'] = df['region'].fillna("United States")
df['device'] = df['device'].fillna("iPhone")
df['session_start'] = df['session_start'].fillna("01.05.2019 2:36")
df['session_end'] = df['session_end'].fillna("01.05.2019 2:45")

In [29]:
df.isna().sum()

user_id          0
region           0
device           0
channel          0
session_start    0
session_end      0
dtype: int64

---
Был выбран способ заполнения пропущенных значений значением по умолчанию, так как удаление пустых строк может привести к ошибкам в дальнейшей работе.
***

### 4. Проверка дубликатов
4.1 Проверка явных дубликатов

In [30]:
df[df.duplicated()]

user_id         region device  channel    session_start  \
480  1,79085E+11  United States    Mac  organic  01.05.2019 7:27   
481  1,79085E+11  United States    Mac  organic  01.05.2019 7:27   

         session_end  
480  01.05.2019 7:49  
481  01.05.2019 7:49

---
Команда `df.duplicated()` проверяет, есть ли дубликаты строк в таблице. Возвращает колонку из True и False
***

In [31]:
df = df.drop_duplicates().reset_index(drop = True)
print(df.duplicated().sum())

0


---
Команда  `df = df.drop_duplicates().reset_index(drop = True)` удаляет дубликаты и обновляет индексы строк
***

4.2 Проверка неявных дубликатов

In [32]:
unique_c = ['region', 'device', 'channel']
for i in unique_c:
  print(i, df[i].unique())

region ['United States' 'USA']
device ['iPhone' 'Mac' 'Android' 'PC' 'MAC' 'IPHONE']
channel ['organic' 'TipTop' 'FaceBoom' 'MediaTornado']


Обнаружены неявные дубликаты: <br>
 `region` - United States : USA; <br>
` device` - 'IPHONE' : 'Iphone'; 'MAC' : 'Mac'.
Дубликаты требуют устранения для дальнейшей обработки данных

In [33]:
df['region'] = df['region'].replace('USA', 'United States')
df['device'] = df['device'].replace('MAC', 'Mac')
df['device'] = df['device'].replace('IPHONE', 'iPhone')
for i in unique_c:
  print(i, df[i].unique())

region ['United States']
device ['iPhone' 'Mac' 'Android' 'PC']
channel ['organic' 'TipTop' 'FaceBoom' 'MediaTornado']


С помощью замены данные приведены к единному виду
***

### 5. Проверка типов данных

In [34]:
df.dtypes

user_id          object
region           object
device           object
channel          object
session_start    object
session_end      object
dtype: object

Проверка показала, что pandas присвоил всем данных тип object, что неправильно, это нужно исправить: <br>
`region`, `device`, `channel`  - переводим в category, так как число записей ограничено; <br>
`session_start и session_end` - переводим в datetime (тип данных для времени).

In [35]:
df['region'] = df['region'].astype('category')
df['device'] = df['device'].astype('category')
df['channel'] = df['channel'].astype('category')

df['session_start'] = pd.to_datetime(df['session_start'], format='%d.%m.%Y %H:%M')
df['session_end'] = pd.to_datetime(df['session_end'], format='%d.%m.%Y %H:%M')

print(df.dtypes)

user_id                  object
region                 category
device                 category
channel                category
session_start    datetime64[ns]
session_end      datetime64[ns]
dtype: object


### 6. Группировка данных
#### Задание 1
Группировка - device и количество рекламных источников каждого типа (channel)

In [36]:
print(df.groupby(['device', 'channel'])['user_id'].count())

device   channel     
Android  FaceBoom         33
         MediaTornado     11
         TipTop           26
         organic         115
Mac      FaceBoom         36
         MediaTornado     10
         TipTop           24
         organic         172
PC       FaceBoom         15
         MediaTornado      2
         TipTop           14
         organic          72
iPhone   FaceBoom         65
         MediaTornado     30
         TipTop           79
         organic         251
Name: user_id, dtype: int64


C:\Users\miros\AppData\Local\Temp\ipykernel_11368\1587863489.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  print(df.groupby(['device', 'channel'])['user_id'].count())



Результаты группировки данных видно, что в изначальном датасете основной источник траффика органический, Iphone доминирует среди устройств, рекламные каналы показывают разную эффективность, аудитория Android меньше Iphone, мобильные устройства используются чаще портативных
***

#### Задание 2

Группировка - region и количество устройств (device ). Создать датафрейм. Переименовать столбец с количеством в “сountˮ. Отсортировать по убыванию столбца “countˮ.

In [37]:
counts = df.groupby(['region', 'device'], observed = True).size()
df_group = counts.reset_index()
df_group.columns = ['region', 'device', 'count']
df_group = df_group.sort_values(by='count', ascending=False)
print(df_group)

          region   device  count
3  United States   iPhone    425
1  United States      Mac    242
0  United States  Android    185
2  United States       PC    103



Вся аудитория из США, самое большое количество пользователей за продуктами Apple (Iphone - 425, Mac - 242)
***

#### Задание 3

In [39]:
pivot_users = df.pivot_table(
    index='device',
    values='user_id',
    aggfunc='nunique',
    observed=True
).sort_values(by='user_id', ascending=False)
print(pivot_users)

         user_id
device          
iPhone       387
Mac          223
Android      169
PC            90




Продукция Apple (Iphone и Mac) - доминирует, продукт нацелен именно на эту систему, нужно улучшать качество продукта для Android и PC и поднимать популярность для пользователей этих устройств.
***

#### Задание 4

Сводная таблица (pivot_table) - уникальное количество пользователей для каждого устройства (device) - столбцы и региона- строки. Отсортировать по возрастанию столбца region.

In [40]:
pivot_users = df.pivot_table(
    index=['region','channel'],
    columns='device',
    values='user_id',
    aggfunc='nunique',
    observed=True
).sort_index(ascending=True)
print(pivot_users)

device                      Android  Mac  PC  iPhone
region        channel                               
United States FaceBoom           32   32  13      63
              MediaTornado       11    9   2      25
              TipTop             19   22  10      69
              organic           108  160  65     232


Из сводной таблицы хорошо видно, что вся аудитория это пользователи из США, явное лидерство у продукции Apple (Iphone и Mac), а платформы Android и PC сильно уступают, что указывает на возможный рост в этой сфере
***

### Вывод

В ходе работы был проведён анализ набора данных visits.csv, содержащего информацию о 957 пользовательских сессиях на цифровой платформе.
**На этапе предобработки были выполнены следующие шаги:**
* Устранены пропуски в столбцах region, device, session_start, session_end
* Преобразован тип данных для столбцов region, device, channel, session_start, session_end

**Выводы из анализа данных с помощью группировок и сводных таблиц:**
* Все пользователи из United States
* Iphone является лидером среди устройтсв, за ним идет Mac, затем Android и PC
* Наибольшее количество уникальных пользователей дублирует картину, что свидетельствует о том, что основная аудитория - пользователи Apple
* Органический траффик - основной источник для всех платформ
***